### PR TITLE
Send 400 Bad Request when HTTP2 cleartext upgrade contains invalid headers

### LINKED ISSUE #4016: Send 400 Bad Request when HTTP2 cleartext ugprade contains an invalid `HTTP2-Settings`
As a follow-up issue mentioned in https://github.com/line/armeria/pull/3913#discussion_r781810461
We should send 400 Bad Request if an upgrade request contains duplicate `HTTP2-Settings`s.
https://github.com/line/armeria/blob/98e7450ce83f7094fb32dc42de7d3251ca1ab962/core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java#L91-L95

### PR DESCRIPTION
**Pull request by ahnyujin**
Motivation:

We should send 400 Bad Request if an upgrade request contains duplicate `HTTP2-Settings`s.
https://github.com/line/armeria/blob/98e7450ce83f7094fb32dc42de7d3251ca1ab962/core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java#L91-L95

Modifications:

- Send 400 Bad Request with  Netty's `FullHttpResponse` if an `upgradeCodec.prepareUpgradeResponse(ctx, request)` is false.

Result:

- Closes #4016

### CODE REVIEW

File: core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java [hunk 1]
```diff
@@ -195,4 +212,87 @@ void upgradeFail() {
         assertThat((Object) channel.readOutbound()).isNull();
         channel.finishAndReleaseAll();
     }
+
+    @Test
+    void upgradeFailWithBadRequest() {
+        final HttpServerCodec httpServerCodec = new HttpServerCodec();
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java:218]: These test suites are previously copied from Netty. I don't see we need to use `UpgradeCodecFactory` to test duplicate `HTTP2-Settings`.
How about 1) starting an Armeria server using `ServerExtension` and 2) sending a raw HTTP/1 upgrade using a JDK's `Socket`?
https://github.com/line/armeria/blob/3491204f94311ab94eb54f224064f14c33896466/core/src/test/java/com/linecorp/armeria/server/ServerMaxConnectionAgeTest.java#L177-L180

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 2]
```diff
@@ -322,4 +326,11 @@ private static FullHttpResponse newUpgradeResponse() {
         res.headers().add(HttpHeaderNames.UPGRADE, Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME);
         return res;
     }
+
+    /**
+     * Creates the 400 Bad Request response message.
+     */
+    private static FullHttpResponse newBadRequestResponse() {
+        return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST, Unpooled.EMPTY_BUFFER);
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:335]: Should we return a 400 response with a detail reason phrase?
```suggestion
    private static FullHttpResponse newInvalidSettingsHeadersResponse() {
        return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST,
                                           Unpooled.copiedBuffer("Invalid HTTP2-Settings header",
                                                                 StandardCharsets.UTF_8));
    }
```

File: core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java [hunk 3]
```diff
@@ -195,4 +211,23 @@ void upgradeFail() {
...
+            writer.print("POST / HTTP/1.1\r\n");
+            writer.print("Content-Length: " + upgradeBody.getBytes(StandardCharsets.UTF_8).length + "\r\n");
+            writer.print("Connection: Upgrade, HTTP2-Settings\r\n");
+            writer.print("Upgrade: h2c\r\n");
+            writer.print("HTTP2-Settings: AAMAAABkAAQAAP__\r\n");
+            writer.print("HTTP2-Settings: AAMAAABkAAQAAP__\r\n\r\n");
+            writer.flush();
+
+            final BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            assertThat(in.readLine()).isEqualTo("HTTP/1.1 400 Bad Request");
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java:230]: 👍

**Review by ikhoon — COMMENTED** [review]
Looking good! Left a minor comment. 🙇‍♂️

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 4]
```diff
@@ -322,4 +328,13 @@ private static FullHttpResponse newUpgradeResponse() {
...
         return res;
     }
+
+    /**
+     * Creates the 400 Bad Request response message.
+     */
+    private static FullHttpResponse newInvalidSettingsHeadersResponse() {
+        return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST,
+                                           Unpooled.copiedBuffer("Invalid HTTP2-Settings headers\r\n",
+                                                                 StandardCharsets.UTF_8));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:338]: Could create a direct buffer and duplicate it every time:

```java
private static ByteBuf INVALID_SETTINGS_HEADERS_MESSAGE =
    Unpooled.directBuffer().writeBytes(...);

...
private static FullHttpResponse newInvalidSettingsHeadersResponse() {
    return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST,
                                       INVALID_SETTINGS_HEADERS_MESSAGE.duplicate());
}
```

File: core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java [hunk 5]
```diff
@@ -195,4 +211,28 @@ void upgradeFail() {
         assertThat((Object) channel.readOutbound()).isNull();
         channel.finishAndReleaseAll();
     }
+
+    @Test
+    void upgradeFailWithInvalidSettingsHeaders() throws Exception {
+        try (Socket socket = new Socket("127.0.0.1", server.httpPort())) {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java:217]: Could you specify the socket timeout, so that this test fails when the server doesn't respond?

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 6]
```diff
@@ -322,4 +328,13 @@ private static FullHttpResponse newUpgradeResponse() {
         res.headers().add(HttpHeaderNames.UPGRADE, Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME);
         return res;
     }
+
+    /**
+     * Creates the 400 Bad Request response message.
+     */
+    private static FullHttpResponse newInvalidSettingsHeadersResponse() {
+        return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST,
+                                           Unpooled.copiedBuffer("Invalid HTTP2-Settings headers\r\n",
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:337]: Should it be headers or header?

**Review by trustin — COMMENTED** [review]
Thanks for the fix, @ahnyujin 🙇

**Review by jrhee17 — APPROVED** [review]
Looks good to me once @trustin 's comments are addressed 👍

**Comment by minwoox** [comment]
Thanks!

**Review by minwoox — APPROVED** [review]

**Comment by trustin** [comment]
Gentle ping, @ahnyujin 🙇

**Comment by ahnyujin** [comment]
> Gentle ping, @ahnyujin 🙇

Thanks for the meaningful comments. I'll try to fix it quickly! 👍

File: core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java [hunk 7]
```diff
@@ -195,4 +211,28 @@ void upgradeFail() {
...
+            writer.print("HTTP2-Settings: AAMAAABkAAQAAP__\r\n\r\n");
+            writer.flush();
+
+            final BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            assertThat(in.readLine()).isEqualTo("HTTP/1.1 400 Bad Request");
+
+            // empty line
+            in.readLine();
+
+            assertThat(in.readLine()).isEqualTo("Invalid HTTP2-Settings headers");
```
[the remark this answers is not in the document]
Comment (ahnyujin) [reply core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java]: That's a really good idea and i've changed it 😀 ! But some question remain. I can't think of any clear reason why the server should not send anything after that.

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 8]
```diff
@@ -276,6 +289,8 @@ private boolean upgrade(ChannelHandlerContext ctx, HttpRequest request) {
         // Prepare and send the upgrade response. Wait for this write to complete before upgrading,
         // since we need the old codec in-place to properly encode the response.
         if (!upgradeCodec.prepareUpgradeResponse(ctx, request)) {
+            ctx.writeAndFlush(INVALID_SETTINGS_HEADER_RESPONSE.retain()).addListener(CLOSE_ON_FAILURE);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:292]: As an invalid/bad request is received, we may close the connection after the write?
```suggestion
            ctx.writeAndFlush(INVALID_SETTINGS_HEADER_RESPONSE.retain()).addListener(CLOSE);
```

**Comment by ikhoon** [comment]
Looks good with the new changes.

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 9]
```diff
@@ -75,6 +79,12 @@ final class HttpServerUpgradeHandler extends ChannelInboundHandlerAdapter {
 
     private static final FullHttpResponse UPGRADE_RESPONSE = newUpgradeResponse();
 
+    private static final ByteBuf INVALID_SETTINGS_HEADER_MESSAGE =
+            Unpooled.directBuffer().writeBytes(
+                    "Invalid HTTP2-Settings header\n".getBytes(StandardCharsets.UTF_8));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:84]: We should make this unreleasable to avoid any potential future mistake, e.g.

```
Unpooled.unreleseableBuffer(Unpooled.directBuffer()....)
```

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 10]
```diff
@@ -75,6 +79,12 @@ final class HttpServerUpgradeHandler extends ChannelInboundHandlerAdapter {
...
+    private static final FullHttpResponse INVALID_SETTINGS_HEADER_RESPONSE =
+            newInvalidSettingsHeaderResponse();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:86]: This should not be cached, because otherwise we'll get `InvalidRefCntException`.

File: core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java [hunk 11]
```diff
@@ -195,4 +211,31 @@ void upgradeFail() {
         assertThat((Object) channel.readOutbound()).isNull();
         channel.finishAndReleaseAll();
     }
+
+    @Test
+    void upgradeFailWithInvalidSettingsHeader() throws Exception {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java:216]: Could we repeat this test twice to make sure we don't have `IllegalRefCntException` or corruption because we didn't duplicate the buffer for each error response? You can use `RepeatedTest` annotation from JUnit.

Comment (minwoox) [reply core/src/test/java/com/linecorp/armeria/server/HttpServerUpgradeHandlerTest.java:216]: @ahnyujin Could you address @trustin's issue, please? We are about to release a new version, so please let me know if you don't have time for this. We are going to fix that. 😄

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 12]
```diff
@@ -75,8 +79,14 @@ final class HttpServerUpgradeHandler extends ChannelInboundHandlerAdapter {
 
     private static final FullHttpResponse UPGRADE_RESPONSE = newUpgradeResponse();
 
+    private static final ByteBuf INVALID_SETTINGS_HEADER_MESSAGE = Unpooled.unreleasableBuffer(
+            Unpooled.directBuffer()
+                    .writeBytes("Invalid HTTP2-Settings header\n".getBytes(StandardCharsets.UTF_8)));
+
     private static final Splitter COMMA_SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
 
+    private final FullHttpResponse invalidSettingsHeaderResponse = newInvalidSettingsHeaderResponse();
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:88]: How about removing this field and creating this response at line 293?
This `invalidSettingsHeaderResponse` is not used if `prepareUpgradeResponse` is `true` and it's used only once when `prepareUpgradeResponse` returns `true`.
We don't need this as a member.

**Review by trustin — APPROVED** [review]
LGTM once @minwoox's comments are addressed.

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java (hunk 12 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:88]: Had a chat with @trustin and let's do:
- make this field static and inline method
```suggestion
    private static final FullHttpResponse invalidSettingsHeaderResponse =
            new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST, INVALID_SETTINGS_HEADER_MESSAGE.duplicate());
```
- Use `duplicate` when it's used. `invalidSettingsHeaderResponse.retain()`

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 13]
```diff
@@ -276,6 +291,8 @@ private boolean upgrade(ChannelHandlerContext ctx, HttpRequest request) {
         // Prepare and send the upgrade response. Wait for this write to complete before upgrading,
         // since we need the old codec in-place to properly encode the response.
         if (!upgradeCodec.prepareUpgradeResponse(ctx, request)) {
+            ctx.writeAndFlush(invalidSettingsHeaderResponse.retain()).addListener(CLOSE);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:294]:
```suggestion
            ctx.writeAndFlush(invalidSettingsHeaderResponse.duplicate()).addListener(CLOSE);
```

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 14]
```diff
@@ -322,4 +339,11 @@ private static FullHttpResponse newUpgradeResponse() {
         res.headers().add(HttpHeaderNames.UPGRADE, Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME);
         return res;
     }
+
+    /**
+     * Creates the 400 Bad Request response message.
+     */
+    private static FullHttpResponse newInvalidSettingsHeaderResponse() {
+        return new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST, INVALID_SETTINGS_HEADER_MESSAGE.duplicate());
+    }
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:348]:
```suggestion
```

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 15]
```diff
@@ -322,4 +339,11 @@ private static FullHttpResponse newUpgradeResponse() {
         res.headers().add(HttpHeaderNames.UPGRADE, Http2CodecUtil.HTTP_UPGRADE_PROTOCOL_NAME);
         return res;
     }
+
+    /**
+     * Creates the 400 Bad Request response message.
+     */
+    private static FullHttpResponse newInvalidSettingsHeaderResponse() {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:346]: Seems not to be used anymore

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java]: I was about to push the commit. 😅

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java]: 😆 I'm OK with the new one.

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 16]
```diff
@@ -75,8 +79,15 @@ final class HttpServerUpgradeHandler extends ChannelInboundHandlerAdapter {
 
     private static final FullHttpResponse UPGRADE_RESPONSE = newUpgradeResponse();
 
+    private static final ByteBuf INVALID_SETTINGS_HEADER_MESSAGE = Unpooled.unreleasableBuffer(
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:82]: Could inline?

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java [hunk 17]
```diff
@@ -75,8 +79,15 @@ final class HttpServerUpgradeHandler extends ChannelInboundHandlerAdapter {
...
+            Unpooled.directBuffer()
+                    .writeBytes("Invalid HTTP2-Settings header\n".getBytes(StandardCharsets.UTF_8)));
+
     private static final Splitter COMMA_SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
 
+    private static final FullHttpResponse invalidSettingsHeaderResponse =
+            new DefaultFullHttpResponse(HTTP_1_1, BAD_REQUEST, INVALID_SETTINGS_HEADER_MESSAGE.duplicate());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:89]: No need to duplicate here because we always duplicate below

**Review by trustin — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java (hunk 17 above)
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/server/HttpServerUpgradeHandler.java:89]: Let me push a commit for that. 😉

**Review by trustin — APPROVED** [review]
Thanks a lot, @ahnyujin and folks 🙇

**Comment by minwoox** [comment]
Thanks a lot, @ahnyujin! 😄

**Comment by ahnyujin** [comment]
Thanks a lot, I learned a lot from your great ideas! 🙇‍♂️
